### PR TITLE
fix(daemon,web): refresh model cache on Gemini OAuth account changes

### PR DESCRIPTION
**Pull request by lsm**
Invalidate the model cache after Gemini OAuth account add/remove/reauth and ensure the frontend refreshes the model list automatically.

**Daemon:**
- Call `clearModelsCache()` in `auth.gemini.completeOAuth` (new account + re-auth paths) and `auth.gemini.removeAccount` so the next `models.list` request re-evaluates provider availability.
- Add an empty-cache fallback in the `models.list` handler: if `getAvailableModels` returns an empty array while `useCache` is true, force a `refreshModels()` before responding.

**Web:**
- `GeminiAccountsPanel` dispatches a `gemini-accounts-changed` window event after successful account add/remove.
- `useModelSwitcher` listens for this event and reloads the model list, so newly available Gemini models appear in the composer without a page refresh.

Fixes Gemini models not appearing in the chat composer after OAuth login.

### CODE REVIEW

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Opus 4.7 (Anthropic)

> **Model:** Claude Opus 4.7 | **Client:** NeoKai | **Provider:** Anthropic

### Summary

The implementation correctly addresses both root causes: stale model cache after OAuth account changes and the frontend's inability to pick up new models. The daemon-side `clearModelsCache()` calls are well-placed (only on success paths), the empty-cache fallback in `models.list` is a clean safety net, and the `gemini-accounts-changed` window event correctly bridges the settings panel to the model switcher.

**One P2 finding** — a missing unit test for the empty-cache fallback in `session-handlers.ts`.

---

### Findings

#### P2: Missing unit test for `models.list` empty-cache fallback

**File:** `packages/daemon/src/lib/rpc-handlers/session-handlers.ts` (lines 711–714)

The new fallback is the critical integration point that ties together the daemon-side cache clearing and the frontend's `models.list` request. When `clearModelsCache()` is called after an OAuth account change, the cache becomes empty. The next `models.list` with `useCache: true` hits this fallback and triggers a forced `refreshModels()`, which re-evaluates all provider availability (including the newly-available Gemini provider).

While the logic is straightforward, it's the linchpin of the fix and should have a dedicated unit test to prevent silent regressions. A future refactor of the `models.list` handler could accidentally remove the fallback or change the condition without realizing its importance.

The existing `session-handlers.test.ts` is an online test (requires API keys), so this path is not covered in CI. A unit test (similar to `gemini-auth-handlers.test.ts` with mocked `model-service` imports) should verify:

1. When `getAvailableModels('global')` returns `[]` and `forceRefresh` is `false`, `refreshModels()` is called and the response contains fresh models.
2. The `cached` response flag is `false` when the fallback fires.
3. The fallback does NOT fire when `forceRefresh` is already `true` (since the forced refresh already ran above).

---

### What looks good

- **`clearModelsCache()` placement**: Only called on success paths (new account, re-auth, remove). All error paths correctly skip it. Verified: re-auth email mismatch, duplicate email, no refresh token, exchange errors, remove-not-found.
- **Test coverage for auth handlers**: Thorough — every success and error path verifies whether `clearModelsCache` was or wasn't called. The retry-after-exchange-error test confirming exactly 1 call is especially nice.
- **Frontend event wiring**: `gemini-accounts-changed` dispatches at exactly the right places (add through modal `onComplete`, remove after success response). The `useModelSwitcher` listener properly cleans up on unmount.
- **`cached` flag update**: Correctly changed to `!forceRefresh && availableModels.length > 0` so the fallback path returns `cached: false`.
- **No over-engineering**: The window CustomEvent approach is appropriately scoped for a same-tab interaction. The daemonHub + StateManager pattern would be overkill here.

File: packages/daemon/src/lib/rpc-handlers/session-handlers.ts [hunk 1]
```diff
@@ -703,7 +703,15 @@ export function setupSessionHandlers(
 				await refreshModels();
 			}
 
-			const availableModels = getAvailableModels('global');
+			let availableModels = getAvailableModels('global');
+
+			// If cache is empty and we're not already forcing refresh, do a forced refresh.
+			// This handles the case where the cache was just cleared (e.g., after OAuth
+			// account changes) so the next models.list call re-evaluates provider availability.
+			if (!forceRefresh && availableModels.length === 0) {
```
Comment (lsm) [inline packages/daemon/src/lib/rpc-handlers/session-handlers.ts:712]: ## 🤖 Review by Claude Opus 4.7 (Anthropic)

> **Model:** Claude Opus 4.7 | **Client:** NeoKai | **Provider:** Anthropic

**P2 — Missing unit test for this empty-cache fallback.**

This is the critical integration point of the fix. When `clearModelsCache()` is called after an OAuth account change, this fallback ensures the next `models.list` with `useCache: true` triggers a forced refresh. Without a unit test, a future refactor could silently break this path.

A unit test (with mocked `model-service` imports) should verify:
1. When `getAvailableModels('global')` returns `[]` and `forceRefresh` is `false`, `refreshModels()` is called and the response contains fresh models.
2. The `cached` flag is `false` when the fallback fires.
3. The fallback does NOT fire when `forceRefresh` is already `true`.

File: packages/daemon/src/lib/rpc-handlers/auth-handlers.ts [hunk 2]
```diff
@@ -462,6 +465,7 @@ export function setupAuthHandlers(messageHub: MessageHub, authManager: AuthManag
 					).getRotationManager();
 					await rm.removeAccount(req.accountId);
 				}
+				clearModelsCache();
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/rpc-handlers/auth-handlers.ts:468]: Fixed in commit 3d719cf4. Updated `refreshModels()` in `model-service.ts` to restore `FALLBACK_MODELS` when the cache is empty and no providers return models, preventing a permanently empty catalog after removing the last Gemini account.

File: packages/daemon/src/lib/rpc-handlers/session-handlers.ts (hunk 1 above)
Comment (lsm) [reply packages/daemon/src/lib/rpc-handlers/session-handlers.ts:712]: Fixed in commit 121633f5. Added dedicated unit tests in `packages/daemon/tests/unit/2-handlers/rpc-handlers/session-handlers.test.ts` covering all three requirements, plus a fix to the `cached` flag so it correctly reports `false` when the fallback refresh fires.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Opus 4.7 (Anthropic)

> **Model:** Claude Opus 4.7 | **Client:** NeoKai | **Provider:** Anthropic

### Summary

The P2 finding has been addressed with a comprehensive new test file and an improved `didRefresh` flag for the `cached` response. The `refreshModels()` fallback model restoration is a good safety net. However, **all 5 new session-handlers tests are failing in CI** due to a mock path mismatch.

---

### Findings

#### P0: All 5 new `session-handlers.test.ts` tests fail in CI — mock path mismatch

**File:** `packages/daemon/tests/unit/2-handlers/rpc-handlers/session-handlers.test.ts` (line 32)

The `mock.module` path uses `.js` extension:
```typescript
mock.module('../../../../src/lib/model-service.js', () => ({ ... }));
```

But `session-handlers.ts` imports `model-service` **without** `.js` (lines 27, 504, 694):
```typescript
import { clearModelsCache } from '../model-service';
const { getAvailableModels, refreshModels } = await import('../model-service');
```

Compare with `auth-handlers.ts` which imports **with** `.js`:
```typescript
import { clearModelsCache } from '../model-service.js';
```

The auth-handlers test mocks `'../../../../src/lib/model-service.js'` and works because the specifier matches. The session-handlers test uses the same `.js` mock path, but the source file's import specifier has no extension — so Bun resolves them differently and the mock doesn't intercept the import.

**Fix:** Change the mock path to match the specifier used in `session-handlers.ts`:
```typescript
mock.module('../../../../src/lib/model-service', () => ({ ... }));
```

**CI error:**
```
SyntaxError: Export named 'getAvailableModels' not found in module '.../model-service.ts'.
error: getAvailableModels is not a function. (In 'getAvailableModels("global")', 'getAvailableModels' is undefined)
```

---

### What looks good (addressed from prior round)

- ✅ **`didRefresh` flag**: Clean improvement over the previous `!forceRefresh` in the `cached` expression. Correctly returns `cached: false` when the empty-cache fallback fires.
- ✅ **`refreshModels()` fallback restoration**: Restores `FALLBACK_MODELS` when cache was empty and no providers return models. Prevents permanently empty catalog.
- ✅ **Test coverage scope**: The 5 test cases cover exactly the right scenarios (cached hit, empty-cache fallback, forceRefresh skip, useCache=false skip, empty-after-refresh).
- ✅ **Model-service tests**: Two new tests for the `refreshModels` fallback restoration behavior.

---

### Unchanged from prior round (still good)

- `clearModelsCache()` placement in auth handlers — only on success paths
- Auth handler test coverage — thorough success/error path verification
- Frontend event wiring — `gemini-accounts-changed` dispatched at correct points
- `useModelSwitcher` event listener with proper cleanup

File: packages/daemon/tests/unit/2-handlers/rpc-handlers/session-handlers.test.ts [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+		contextWindow: number;
+		description: string;
+		releaseDate: string;
+		available: boolean;
+	}> => []
+);
+
+const mockRefreshModels = mock(async () => {});
+
+mock.module('../../../../src/lib/model-service.js', () => ({
```
Comment (lsm) [inline packages/daemon/tests/unit/2-handlers/rpc-handlers/session-handlers.test.ts]: ## 🤖 Review by Claude Opus 4.7 (Anthropic)

> **Model:** Claude Opus 4.7 | **Client:** NeoKai | **Provider:** Anthropic

**P0 — Mock path mismatch causes all 5 tests to fail in CI.**

The mock uses `'../../../../src/lib/model-service.js'` (with `.js`), but this file imports from `'../model-service'` (without `.js`) — see lines 27, 504, 694 of session-handlers.ts. Bun resolves these to different paths, so the mock never intercepts the import.

Fix: change the mock path to match the specifier used in session-handlers.ts:
```typescript
mock.module('../../../../src/lib/model-service', () => ({
    getAvailableModels: mockGetAvailableModels,
    refreshModels: mockRefreshModels,
    clearModelsCache: mock(() => {}),
}));
```

Compare: auth-handlers.ts uses `'../model-service.js'` (with `.js`) so its test mock with `.js` works. session-handlers.ts uses no extension.

File: packages/daemon/src/lib/rpc-handlers/auth-handlers.ts [hunk 4]
```diff
@@ -422,6 +424,7 @@ export function setupAuthHandlers(messageHub: MessageHub, authManager: AuthManag
 				await persistAddAccount(account);
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/rpc-handlers/auth-handlers.ts:424]: Fixed in commit 86060979. Added a `cacheGeneration` counter to `model-service.ts`:
- `clearModelsCache()` bumps the generation and clears `refreshInProgress`
- `triggerBackgroundRefresh()` captures the generation at start and only writes if it matches
This prevents any in-flight background refresh from overwriting the cache with stale pre-change data. Added a unit test verifying the cancellation behavior.

File: packages/daemon/tests/unit/2-handlers/rpc-handlers/session-handlers.test.ts (hunk 3 above)
Comment (lsm) [reply packages/daemon/tests/unit/2-handlers/rpc-handlers/session-handlers.test.ts]: Fixed in commit 86060979. Changed the mock path from `../../../../src/lib/model-service.js` to `../../../../src/lib/model-service` (no extension) to match the import specifier used by `session-handlers.ts`.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Opus 4.7 (Anthropic)

> **Model:** Claude Opus 4.7 | **Client:** NeoKai | **Provider:** Anthropic

### Summary

The P0 mock path fix was applied but **CI still fails** — all 5 `session-handlers.test.ts` tests produce the same `getAvailableModels is not a function` error. The `.js` removal alone doesn't resolve the underlying mock interception issue. Additionally, the new `cacheGeneration` test doesn't exercise the generation counter (findings below).

---

### Findings

#### P0: `session-handlers.test.ts` — all 5 tests still fail in CI

**File:** `packages/daemon/tests/unit/2-handlers/rpc-handlers/session-handlers.test.ts` (line 32)

Removing `.js` from the mock path (`'../../../../src/lib/model-service'` → no extension) did not fix the CI failure. The same error persists:

```
SyntaxError: Export named 'getAvailableModels' not found in module '.../model-service.ts'.
error: getAvailableModels is not a function. (In 'getAvailableModels("global")', 'getAvailableModels' is undefined)
```

**Root cause hypothesis:** The `2-handlers` shard runs 21 test files in a single process. `gemini-auth-handlers.test.ts` (which runs before this file) mocks `'model-service.js'` (with `.js`) and then calls `mock.restore()` in its `afterEach`. When `session-handlers.test.ts` subsequently mocks `'model-service'` (without `.js`), Bun may not match the specifier to the already-cached module loaded under the `.js` key by the earlier test file.

**Suggested fix approaches (in order of preference):**

1. **Don't mock `model-service` at all.** Instead, use the real functions: call `clearModelsCache()` to empty the cache and register a test provider via `getProviderRegistry()` to control what `refreshModels()` returns. This tests actual integration rather than mocked behavior.

2. **Mock only `loadModelsFromProviders`** (the internal function that fetches from providers) instead of the entire `model-service` module. This avoids the cross-file mock specifier conflict.

3. **Match the exact specifier used by `session-handlers.ts`.** Change all 3 occurrences in `session-handlers.ts` (lines 27, 504, 694) from `'../model-service'` to `'../model-service.js'`, then use the `.js` mock path. This ensures consistency with the rest of the codebase.

---

#### P2: `cacheGeneration` test doesn't exercise the generation counter

**File:** `packages/daemon/tests/unit/1-core/core/model-service.test.ts` (line 1105)

The test "should cancel in-flight background refresh when clearModelsCache is called" validates the correct outcome (cache stays empty), but doesn't actually exercise the `cacheGeneration` guard. Because no providers are registered in the test environment, `loadModelsFromProviders()` returns `[]`, and the `models.length > 0` check at line 221 fails before the generation counter is evaluated. The test would pass identically without the generation mechanism.

To properly test the generation counter, register a mock provider that returns models after an artificial delay, so the background refresh has data that the generation counter must then suppress:

```typescript
// Register a slow provider that returns models
const mockProvider = {
    isAvailable: () => true,
    getModels: async () => {
        await new Promise(r => setTimeout(r, 100));
        return [/* some models */];
    },
};
getProviderRegistry().register(mockProvider);
```

---

### What looks good

- ✅ **`cacheGeneration` mechanism**: Clean design. The monotonic counter in `triggerBackgroundRefresh` correctly prevents stale background writes after `clearModelsCache()`. The `refreshInProgress` map clearing in `clearModelsCache()` is correctly idempotent with the `finally` block.
- ✅ **`refreshModels()` not using generation counter**: Correct by design — it's the synchronous forced refresh path that re-fetches from providers, so it bypasses the generation guard.
- ✅ **FALLBACK_MODELS restoration**: Restores baseline catalog when cache is empty and no providers return models.
- ✅ **`didRefresh` flag**: Correctly tracks whether any refresh happened for the `cached` response flag.
- ✅ **Auth handler `clearModelsCache` calls**: Unchanged from prior rounds, still correct.
- ✅ **Frontend event wiring**: Unchanged, still correct.

File: packages/daemon/tests/unit/2-handlers/rpc-handlers/session-handlers.test.ts [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+		contextWindow: number;
+		description: string;
+		releaseDate: string;
+		available: boolean;
+	}> => []
+);
+
+const mockRefreshModels = mock(async () => {});
+
+mock.module('../../../../src/lib/model-service', () => ({
```
Comment (lsm) [inline packages/daemon/tests/unit/2-handlers/rpc-handlers/session-handlers.test.ts]: ## 🤖 Review by Claude Opus 4.7 (Anthropic)

> **Model:** Claude Opus 4.7 | **Client:** NeoKai | **Provider:** Anthropic

**P0 — Mock path fix didn't resolve CI failure. All 5 tests still fail.**

The `.js` removal alone didn't fix the issue. The same `getAvailableModels is not a function` error persists in CI.

Root cause: The `2-handlers` shard runs 21 test files in a single Bun process. `gemini-auth-handlers.test.ts` (which runs earlier) mocks `model-service.js` (with `.js`) and calls `mock.restore()` in `afterEach`. When this test subsequently mocks `model-service` (without `.js`), Bun likely doesn't match the specifier to the already-cached module loaded under the `.js` key.

**Suggested fixes (in order of preference):**

1. **Don't mock `model-service` at all.** Use real functions: call `clearModelsCache()` to empty the cache, register a test provider via `getProviderRegistry()` to control what `refreshModels()` returns. Tests actual integration.

2. **Mock only `loadModelsFromProviders`** instead of the entire module.

3. **Change `session-handlers.ts` imports to use `.js`** extension (lines 27, 504, 694: `'../model-service'` → `'../model-service.js'`), then use the `.js` mock path. Consistent with `auth-handlers.ts`.

File: packages/daemon/src/lib/model-service.ts [hunk 6]
```diff
@@ -335,14 +347,21 @@ function clearProviderModelCaches(): void {
 /**
  * Clear the models cache for a specific key or all.
  * Provider-level caches are only cleared on a full clear (no cacheKey).
+ *
+ * Also bumps the cache generation so any in-flight background refreshes
+ * that started before this call drop their stale results instead of
+ * overwriting the cleared cache.
  */
 export function clearModelsCache(cacheKey?: string): void {
+	cacheGeneration++;
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/model-service.ts]: Fixed in commit 4a15b636. Replaced the single global cacheGeneration counter with a Map keyed by cache key. clearModelsCache(session-123) now only bumps the generation for that key, so it can no longer cancel an in-flight triggerBackgroundRefresh(global). Global clears (no key) still bump every known key plus global to preserve the existing safety net. Added a unit test verifying that a session-scoped clear does not abort a global background refresh.

File: packages/daemon/tests/unit/2-handlers/rpc-handlers/session-handlers.test.ts (hunk 5 above)
Comment (lsm) [reply packages/daemon/tests/unit/2-handlers/rpc-handlers/session-handlers.test.ts]: Fixed in commit 77c69e20. Changed session-handlers.ts to import model-service with .js extension (consistent with auth-handlers.ts) and updated the test mock path to match. This ensures both files load the same module instance in the 2-handlers shard, avoiding the Bun specifier mismatch.

File: packages/daemon/src/lib/model-service.ts [hunk 7]
```diff
@@ -335,22 +348,39 @@ function clearProviderModelCaches(): void {
...
+ * Also bumps the per-key cache generation so any in-flight background
+ * refreshes for that key drop their stale results instead of overwriting
+ * the cleared cache.
  */
 export function clearModelsCache(cacheKey?: string): void {
 	if (cacheKey) {
 		modelsCache.delete(cacheKey);
 		cacheTimestamps.delete(cacheKey);
+		refreshInProgress.delete(cacheKey);
+		cacheGeneration.set(cacheKey, (cacheGeneration.get(cacheKey) ?? 0) + 1);
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/model-service.ts]: Fixed in commit 77c69e20. clearModelsCache now deletes generation tracking for keys with no in-flight refresh, preventing unbounded growth of session IDs. triggerBackgroundRefresh also cleans up generation keys in its finally block when the cache entry has been removed. Global clears bump generation only for in-flight keys and prune the rest.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Opus 4.7 (Anthropic)

> **Model:** Claude Opus 4.7 | **Client:** NeoKai | **Provider:** Anthropic

### Summary

The per-key `cacheGeneration` Map is a solid improvement and the new tests are well-designed. However, **CI still fails** for `session-handlers.test.ts` — all 5 tests produce the same `getAvailableModels is not a function` error. The `.js` import alignment didn't fix it. The root cause is Bun's `mock.module` behavior when multiple test files in the same shard mock the same module.

---

### Findings

#### P0: `session-handlers.test.ts` — CI failure persists (3rd round)

The mock path and import specifier alignment (`'.js'` everywhere) didn't resolve the CI failure. Same error:

```
SyntaxError: Export named 'getAvailableModels' not found in module '.../model-service.ts'.
```

**Root cause:** In the `2-handlers` shard, `gemini-auth-handlers.test.ts` runs first and loads `model-service.js` (through `auth-handlers.ts`'s static import). When `mock.restore()` is called in its `afterEach`, the module is restored but stays cached. When `session-handlers.test.ts` subsequently registers its mock, Bun doesn't re-apply it to the already-cached module — even though the specifier now matches.

**Recommended fix: Drop the `mock.module` approach entirely.** Use real `model-service` functions with controlled cache state. The handler's empty-cache fallback logic can be fully tested without mocking:

```typescript
import { clearModelsCache, setModelsCache, getAvailableModels } from '../../../../src/lib/model-service';

beforeEach(() => {
    clearModelsCache();
});

it('returns cached models when cache is populated', async () => {
    // Seed cache, call handler, verify cached: true, no refresh
    const testCache = new Map<string, ModelInfo[]>();
    testCache.set('global', [/* mock model */]);
    setModelsCache(testCache);
    const result = await handler({ useCache: true });
    expect(result.cached).toBe(true);
});

it('triggers refresh when cache is empty', async () => {
    // Cache cleared, call handler, verify refreshModels runs (fallback models restored)
    const result = await handler({ useCache: true });
    expect(result.cached).toBe(false);
    // After refreshModels, FALLBACK_MODELS are restored
    expect(result.models.length).toBeGreaterThan(0);
});
```

This tests actual integration rather than mocked behavior, and avoids all `mock.module` cross-file issues. Note: the "returns empty array when cache stays empty after refresh" test case can be dropped — after the `refreshModels()` fallback restoration fix, the cache is never permanently empty.

---

### What looks good (new in this round)

- ✅ **Per-key `cacheGeneration` Map**: Clean improvement over the global counter. Session-scoped clears no longer cancel global refreshes.
- ✅ **Generation key pruning**: Properly cleans up per-session generation entries to prevent unbounded growth.
- ✅ **Background refresh tests with real providers**: The `model-service.test.ts` tests now register actual mock providers, properly exercising the generation counter.
- ✅ **Session-scoped clear isolation test**: New test verifies `clearModelsCache('session-123')` doesn't cancel a background refresh for a different key.
- ✅ **Custom cache keys in tests**: Avoids cross-test async contamination on `'global'`.

---

### Unchanged from prior rounds (still correct)

- Auth handler `clearModelsCache()` calls on success paths only
- Auth handler test coverage — thorough success/error path verification
- Frontend event wiring and `useModelSwitcher` listener
- `didRefresh` flag and `cached` response logic
- FALLBACK_MODELS restoration in `refreshModels()`
- Empty-cache fallback in `models.list` handler

File: packages/daemon/tests/unit/2-handlers/rpc-handlers/session-handlers.test.ts (hunk 3 above)
Comment (lsm) [inline packages/daemon/tests/unit/2-handlers/rpc-handlers/session-handlers.test.ts]: ## 🤖 Review by Claude Opus 4.7 (Anthropic)

> **Model:** Claude Opus 4.7 | **Client:** NeoKai | **Provider:** Anthropic

**P0 — CI failure persists (3rd round). Mock.module approach fundamentally doesn't work for this file in the 2-handlers shard.**

Same error despite `.js` alignment. Root cause: `gemini-auth-handlers.test.ts` runs first in the shard, loads `model-service.js` into Bun's module cache, then `mock.restore()` removes the mock but the real module stays cached. When this test's `mock.module` runs, Bun doesn't re-apply it to the already-cached module.

**Fix: Drop `mock.module` entirely.** Use real `model-service` functions with controlled cache state:

```typescript
import { clearModelsCache, setModelsCache } from '../../../../src/lib/model-service';

beforeEach(() => { clearModelsCache(); });

it('returns cached models when cache is populated', async () => {
    const testCache = new Map<string, ModelInfo[]>();
    testCache.set('global', [/* mock model */]);
    setModelsCache(testCache);
    const result = await handler({ useCache: true });
    expect(result.cached).toBe(true);
});
```

This tests actual integration and avoids all `mock.module` cross-file issues. The "returns empty array when cache stays empty" test case can be dropped — after the `refreshModels()` fallback restoration fix, the cache is never permanently empty.

File: packages/daemon/src/lib/model-service.ts [hunk 8]
```diff
@@ -335,22 +352,52 @@ function clearProviderModelCaches(): void {
...
+		refreshInProgress.clear();
 		clearProviderModelCaches();
+		// Bump generation only for keys that had in-flight refreshes.
+		for (const key of inFlightKeys) {
+			cacheGeneration.set(key, (cacheGeneration.get(key) ?? 0) + 1);
+		}
+		// Prune generation tracking for keys with no in-flight refresh.
+		for (const key of Array.from(cacheGeneration.keys())) {
+			if (!inFlightKeys.has(key)) {
+				cacheGeneration.delete(key);
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/model-service.ts]: Fixed in commit 900b3fa. Removed the pruning loop from the global clearModelsCache path. After a first global clear bumps a generation to invalidate an in-flight refresh, a second rapid clear no longer deletes that bump, so the stale refresh still drops its result instead of repopulating the cache.

File: packages/daemon/tests/unit/2-handlers/rpc-handlers/session-handlers.test.ts (hunk 3 above)
Comment (lsm) [reply packages/daemon/tests/unit/2-handlers/rpc-handlers/session-handlers.test.ts]: Fixed in commit e218299e. Rewrote session-handlers.test.ts to drop mock.module entirely and use real model-service functions with controlled cache state. Discovered that gemini-auth-handlers.test.ts installs a top-level mock.module on model-service.js that Bun does not fully restore, leaving clearModelsCache as a no-op for the rest of the shard. We now use setModelsCache(new Map()) to empty the cache (setModelsCache is unaffected by the mock) and resetProviderRegistry/resetProviderFactory for provider isolation. All 4 session-handlers tests pass alongside the rest of the 2-handlers shard.

File: packages/daemon/src/lib/model-service.ts [hunk 9]
```diff
@@ -335,22 +352,54 @@ function clearProviderModelCaches(): void {
...
 export function clearModelsCache(cacheKey?: string): void {
 	if (cacheKey) {
+		const hadInFlight = refreshInProgress.has(cacheKey);
 		modelsCache.delete(cacheKey);
 		cacheTimestamps.delete(cacheKey);
+		refreshInProgress.delete(cacheKey);
+		if (hadInFlight) {
+			// Bump so the in-flight refresh drops its stale result.
+			cacheGeneration.set(cacheKey, (cacheGeneration.get(cacheKey) ?? 0) + 1);
+		} else {
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/model-service.ts]: Fixed in commit e218299e. clearModelsCache(cacheKey) now bumps the generation whenever the key has an in-flight refresh OR existing generation history. A second rapid clear can no longer delete the bump because the cacheGeneration.has(cacheKey) guard preserves it. The stale refresh sees a mismatched generation and drops its result.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Opus 4.7 (Anthropic)

> **Model:** Claude Opus 4.7 | **Client:** NeoKai | **Provider:** Anthropic

### Verdict: APPROVE ✅

All prior findings addressed. CI is green across the board. All review threads resolved.

---

### Summary of final state

**Daemon side:**
- `clearModelsCache()` called on all 3 OAuth success paths (add, re-auth, remove) in `auth-handlers.ts`
- Empty-cache fallback in `models.list` handler forces `refreshModels()` when cache is empty and `useCache=true`
- `didRefresh` flag correctly reports `cached: false` when any refresh occurred
- Per-key `cacheGeneration` Map prevents stale background refresh writes after cache clears
- Generation key pruning prevents unbounded growth; global clears don't prune (correctly prevents double-clear race)
- `FALLBACK_MODELS` restoration ensures the model catalog is never permanently empty
- All `session-handlers.ts` imports aligned to `.js` extension for consistency

**Frontend side:**
- `GeminiAccountsPanel` dispatches `gemini-accounts-changed` after successful add/remove
- `useModelSwitcher` listens for the event and reloads model list
- Proper event listener cleanup on unmount

**Tests:**
- Auth handler tests: 17/17 — thorough coverage of `clearModelsCache` on success paths and no-call on error paths
- Session handler tests: 4/4 — uses real `model-service` functions with `setModelsCache` state control, avoiding `mock.module` cross-file contamination
- Model-service tests: 4 new tests for `refreshModels` fallback restoration, background refresh cancellation, per-key generation isolation, and session-scoped clear isolation
- Frontend tests: `useModelSwitcher` event listener test, `GeminiAccountsPanel` event dispatch tests for add and remove

**CI:** All checks pass (lint, web tests, daemon unit tests 1-core + 2-handlers).

File: packages/daemon/src/lib/model-service.ts [hunk 10]
```diff
@@ -335,22 +352,53 @@ function clearProviderModelCaches(): void {
...
+			// including one invalidated by an earlier clear — drops its result.
+			cacheGeneration.set(cacheKey, (cacheGeneration.get(cacheKey) ?? 0) + 1);
+		}
+		// If there's no in-flight refresh and no generation history, there's
+		// nothing to invalidate; leave the key absent.
 	} else {
+		const inFlightKeys = new Set(refreshInProgress.keys());
 		modelsCache.clear();
 		cacheTimestamps.clear();
+		refreshInProgress.clear();
```
[the remark this answers is not in the document]
Comment (lsm) [reply packages/daemon/src/lib/model-service.ts:377]: Fixed. `refreshModels()` now tracks itself in `refreshInProgress`, captures a `generationAtStart` snapshot, and checks the generation before writing. If `clearModelsCache()` is called while a foreground refresh is running, the generation mismatch causes the refresh to drop its stale result instead of overwriting the cache. Added a unit test with a slow provider that verifies the stale result is dropped.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Opus 4.7 (Anthropic)

> **Model:** Claude Opus 4.7 | **Client:** NeoKai | **Provider:** Anthropic

### Verdict: APPROVE ✅

All prior findings addressed. The new foreground refresh generation guard is a clean extension of the existing pattern. CI green across the board.

---

### New change review: `refreshModels()` generation guard

The latest commit extends the generation counter protection to the foreground (`refreshModels()`) path, not just the background (`triggerBackgroundRefresh`) path. This is the correct design — without it, `clearModelsCache()` during a foreground refresh wouldn't prevent the stale result from being written.

**What changed in `refreshModels()`:**
- Now registers itself in `refreshInProgress` (enabling de-duplication and `clearModelsCache()` detection)
- Captures `generationAtStart` before async work, checks it before writing
- Drops results if generation changed (i.e., `clearModelsCache()` was called during the fetch)
- `finally` block cleans up `refreshInProgress` and prunes generation tracking

**New test:** Registers a slow provider (200ms delay), clears cache mid-refresh, verifies stale result is dropped. Properly exercises the generation guard in the foreground path. Test runs in the 1-core shard where `clearModelsCache()` is not mocked.

**CI:** All checks pass on the latest run, including `Daemon Online (git)` (the failure on the first run was a pre-existing flaky test unrelated to this PR).
